File: Leetcode/linkedlist/1019_NextGreaterNodeInLinkedList/Solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def nextLargerNodes(self, head):
        """
        :type head: ListNode
        :rtype: List[int]
        """
        stk = []
        nums = []
        temp = head
        while temp:
            nums.append(temp.val)
            temp = temp.next
        res = [0]*len(nums)
        for i in range(len(nums)):
            while stk and nums[stk[-1]]<nums[i]:
                res[stk.pop()] = nums[i]
            stk.append(i)
        return res

File: Leetcode/linkedlist/1019_NextGreaterNodeInLinkedList/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public int[] nextLargerNodes(ListNode head) {
        Stack<Integer> stk = new Stack<>();
        ListNode temp = head;
        List<Integer> nums = new ArrayList<>();
        while(temp!=null){
            nums.add(temp.val);
            temp = temp.next;
        }
        int[] res = new int[nums.size()];
        for(int i = 0; i<nums.size(); ++i){
            while(!stk.isEmpty() && nums.get(stk.peek())<nums.get(i)){
                int index = stk.pop();
                res[index] = nums.get(i);
            }
            stk.push(i);
        }
        return res;
    }
}
